File: treino/exercicios beewcorwd/4-ESTRUTURAS E BIBLIOTECAS/camisetas.java
import java.util.Scanner;

public class camisetas {

    static class No {
        String nome, cor;
        char tamanho;
        No esq, dir;

        No(String nome, String cor, char tamanho) {
            this.nome    = nome;
            this.cor     = cor;
            this.tamanho = tamanho;
            this.esq     = null;
            this.dir     = null;
        }
    }

    static int compareStr(String a, String b) {
        int len = a.length() < b.length() ? a.length() : b.length();
        for (int i = 0; i < len; i++) {
            if (a.charAt(i) != b.charAt(i)) return a.charAt(i) - b.charAt(i);
        }
        return a.length() - b.length();
    }

    // cor ascendente, tamanho descendente (P>M>G), nome ascendente
    static int compare(String corA, char tamA, String nomeA,
                       String corB, char tamB, String nomeB) {
        int c = compareStr(corA, corB);
        if (c != 0) return c;
        c = tamB - tamA; // descendente: inverte a e b
        if (c != 0) return c;
        return compareStr(nomeA, nomeB);
    }

    static No inserir(No raiz, String nome, String cor, char tamanho) {
        if (raiz == null) return new No(nome, cor, tamanho);
        int cmp = compare(cor, tamanho, nome, raiz.cor, raiz.tamanho, raiz.nome);
        if (cmp <= 0) raiz.esq = inserir(raiz.esq, nome, cor, tamanho);
        else          raiz.dir = inserir(raiz.dir, nome, cor, tamanho);
        return raiz;
    }

    static void emOrdem(No raiz, StringBuilder sb) {
        if (raiz == null) return;
        emOrdem(raiz.esq, sb);
        sb.append(raiz.cor).append(" ").append(raiz.tamanho).append(" ").append(raiz.nome).append("\n");
        emOrdem(raiz.dir, sb);
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        StringBuilder sb = new StringBuilder();
        boolean primeiro = true;

        while (sc.hasNextInt()) {
            int n = sc.nextInt();
            if (n == 0) break;
            sc.nextLine(); // consome o restante da linha do N

            No raiz = null;

            for (int i = 0; i < n; i++) {
                String nome    = sc.nextLine();
                String cor     = sc.next();
                char   tamanho = sc.next().charAt(0);
                sc.nextLine(); // consome o restante da linha
                raiz = inserir(raiz, nome, cor, tamanho);
            }

            if (!primeiro) sb.append("\n");
            primeiro = false;

            emOrdem(raiz, sb);
        }

        System.out.print(sb);
    }
}

File: treino/exercicios beewcorwd/4-ESTRUTURAS E BIBLIOTECAS/camisetas.c
#include <stdio.h>

#define MAX_NODES 10001
#define MAX_LEN   201

/* BST representada por arrays paralelos (pool de nós sem malloc) */
char pNome[MAX_NODES][MAX_LEN];
char pCor [MAX_NODES][MAX_LEN];
char pTam [MAX_NODES];
int  pEsq [MAX_NODES];
int  pDir [MAX_NODES];
int  poolCount;

int compareStr(char* a, char* b) {
    int i = 0;
    while (a[i] && b[i]) {
        if (a[i] != b[i]) return (unsigned char)a[i] - (unsigned char)b[i];
        i++;
    }
    if (!a[i] && !b[i]) return 0;
    return a[i] ? 1 : -1;
}

/* cor ASC, tamanho DESC (P>M>G = menor char ganha), nome ASC */
int compare(char* corA, char tamA, char* nomeA,
            char* corB, char tamB, char* nomeB) {
    int c = compareStr(corA, corB);
    if (c != 0) return c;
    c = tamB - tamA; /* invertido: B - A = descendente */
    if (c != 0) return c;
    return compareStr(nomeA, nomeB);
}

void copyStr(char* dst, char* src) {
    int i = 0;
    while (src[i]) { dst[i] = src[i]; i++; }
    dst[i] = '\0';
}

int inserir(int raiz, char* nome, char* cor, char tamanho) {
    if (raiz == -1) {
        int idx = poolCount++;
        copyStr(pNome[idx], nome);
        copyStr(pCor[idx],  cor);
        pTam[idx] = tamanho;
        pEsq[idx] = -1;
        pDir[idx] = -1;
        return idx;
    }
    int cmp = compare(cor, tamanho, nome,
                      pCor[raiz], pTam[raiz], pNome[raiz]);
    if (cmp <= 0) pEsq[raiz] = inserir(pEsq[raiz], nome, cor, tamanho);
    else          pDir[raiz] = inserir(pDir[raiz], nome, cor, tamanho);
    return raiz;
}

void emOrdem(int raiz) {
    if (raiz == -1) return;
    emOrdem(pEsq[raiz]);
    printf("%s %c %s\n", pCor[raiz], pTam[raiz], pNome[raiz]);
    emOrdem(pDir[raiz]);
}

int main() {
    int n, i, raiz, primeiro = 1;
    char nome[MAX_LEN], cor[MAX_LEN], tam[4];
    int c, ch;

    while (scanf("%d", &n) == 1 && n != 0) {
        poolCount = 0;
        raiz = -1;

        /* consome o resto da linha do n */
        while ((ch = getchar()) != '\n' && ch != EOF);

        for (i = 0; i < n; i++) {
            /* lê o nome — pode conter espaços */
            int j = 0;
            while ((c = getchar()) != '\n' && c != EOF)
                if (c != '\r') nome[j++] = c;
            nome[j] = '\0';

            /* lê cor e tamanho */
            scanf("%s %s", cor, tam);

            /* consome o resto da linha do tamanho */
            while ((ch = getchar()) != '\n' && ch != EOF);

            raiz = inserir(raiz, nome, cor, tam[0]);
        }

        if (!primeiro) printf("\n");
        primeiro = 0;

        emOrdem(raiz);
    }

    return 0;
}
